Handling grouped selectors with :matches

johno opened the pull request:
Firstly, thanks for this plugin. It's awesome.

I've encountered an issue using `rework-matches` with grouped CSS selectors. When using a grouped selector like `.foo:matches(:hover), .bar:matches(:hover)` the output gets muddied up.

The issue arises from the fact that rework splits all selectors on commas, and you're doing some awesome magic to work around that shortcoming. In attempt to find a work around, I've introduced a failing test case that can be summarized as:

``` css
.foo:matches(:hover, :active, :focus),
.baz:matches(:hover, :active, :focus) {
  background: red;
}
```

Results in:

``` css
.foo:hover&.baz:hover,
.foo:hover&.baz:active,
.foo:hover&.baz:focus,
.foo:active&.baz:hover,
.foo:active&.baz:active,
.foo:active&.baz:focus,
.foo:focus&.baz:hover,
.foo:focus&.baz:active,
.foo:focus&.baz:focus
```

I figured I'd use this testcase as a starting point to implement a fix. It looks like the way we can discern whether the selector is part of a group or not is when we end up with a comma placeholder token, `&` in the string that's not within the `:matches` parens.

I should have a PR ready to go sometime tomorrow with a fix.

Thanks!

johno committed:
Add failing test

michelle commented:
Thanks for the detailed bug report! :) Looking forward to your patch!

Michelle

On Tue, Feb 3, 2015 at 9:51 PM, John Otander notifications@github.com
wrote:

> Firstly, thanks for this plugin. It's awesome.
> 
> I've encountered an issue using rework-matches with grouped CSS
> selectors. When using a grouped selector like .foo:matches(:hover),
> .bar:matches(:hover) the output gets muddied up.
> 
> The issue arises from the fact that rework splits all selectors on commas,
> and you're doing some awesome magic to work around that shortcoming. In
> attempt to find a work around, I've introduced a failing test case that can
> be summarized as:
> 
> .foo:matches(:hover, :active, :focus),.baz:matches(:hover, :active, :focus) {
>   background: red;
> }
> 
> Results in:
> 
> .foo:hover&.baz:hover,.foo:hover&.baz:active,.foo:hover&.baz:focus,.foo:active&.baz:hover,.foo:active&.baz:active,.foo:active&.baz:focus,.foo:focus&.baz:hover,.foo:focus&.baz:active,.foo:focus&.baz:focus
> 
> I figured I'd use this testcase as a starting point to implement a fix. It
> looks like the way we can discern whether the selector is part of a group
> or not is when we end up with a comma placeholder token, & in the string
> that's not within the :matches parens.
> 
> I should have a PR ready to go sometime tomorrow with a fix.
> 
> ## Thanks!
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/michelle/rework-matches/pull/1
> Commit Summary
> - Add failing test
> 
> File Changes
> - _M_ test/css/matches.css
>   https://github.com/michelle/rework-matches/pull/1/files#diff-0 (3)
> - _M_ test/css/matches.out.css
>   https://github.com/michelle/rework-matches/pull/1/files#diff-1 (5)
> 
> Patch Links:
> - https://github.com/michelle/rework-matches/pull/1.patch
> - https://github.com/michelle/rework-matches/pull/1.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/michelle/rework-matches/pull/1.

johno committed:
Make the selector splitting regex lazy

This will ensure that once a .foo:matches(:hover, :focus, :active)
is hit, it won't continue searching to match a
.baz:matches(:hover, :focus, :active) selector that is grouped.

For example, previously the following:

```
.foo:matches(:hover, :active, :focus),
.baz:matches(:hover, :active, :focus) {
  background: red;
}

.bar :matches(div, p) :matches(.baz, .foo) {
  background: blue;
}
```

Resulted in the following matches:

```
'.foo:matches(:hover&:active&:focus)&.baz:matches(:hover&:active&:focus)',
'.bar :matches(div&p) :matches(.baz&.foo)'
```

After the lazy conversion we result in:

```
'.foo:matches(:hover&:active&:focus)',
'.baz:matches(:hover&:active&:focus)',
'.bar :matches(div&p) :matches(.baz&.foo)'
```

johno committed:
Filter :matches selectors at the end, add tests

johno commented:
Not a problem!

I've incorporated a fix and added a few new test cases. The main change is filtering the `:matches` selectors after the selectors are translated to CSS3 [https://github.com/johnotander/rework-matches/commit/0e859ef911e9596b0581cac2d8a380b77fd9a83a]. This was required for grouped `:matches` because the `splice` caused the second `:matches` selector to be skipped.

Let me know if you'd like anything changed. Thanks!

michelle commented on index.js:
```diff
@@ -34,9 +38,10 @@ module.exports = function() {
       // Unfortunately rework unconditionally splits selectors by comma, so our
       // :matches(...) selector gets munged.
-      var actualSelectors = rule.selectors.join(',').replace(/:matches\(.*\)/g, function(substr) {
+      var actualSelectors = rule.selectors.join(',').replace(/:matches\(.*?\)/g, function(substr) {
```
Sorry, not super familiar with what the combination of `*` and `?` is doing--is the `.` supposed to be both greedy and not?

johno replied:
No worries. I very well could be doing it improperly.

My intention was that the regex will now only match up to the first closing paren. However, the `.*` is greedy and will match everything between `:matches(` and the last `)` of a string. So the regex would continue when there was a chained selector all the way to the last closing paren of the joined selectors:

``` js
$ node
> var joinedSelector = '.foo:matches(:hover,:active,:focus),.baz:matches(:hover,:active,:focus)'
undefined
> joinedSelector.replace(/:matches\(.*\)/g, function(substr) {
  return substr.replace(/,/g, '&')
}).split(',') 
[ '.foo:matches(:hover&:active&:focus)&.baz:matches(:hover&:active&:focus)' ]
> joinedSelector.replace(/:matches\(.*\)/g, function(substr) {
  console.log(substr)
  return substr.replace(/,/g, '&')
}).split(',')
:matches(:hover,:active,:focus),.baz:matches(:hover,:active,:focus)
[ '.foo:matches(:hover&:active&:focus)&.baz:matches(:hover&:active&:focus)' ]
```

So, the `.*` was the right approach, up until the closing paren because it resulted in an `&` outside the `:matches` parens where a selector was originally grouped.

However, according to the article, I could introduce the minimal matcher, `?`, to make the `.*` "lazy but helpful" [[reference](http://www.rexegg.com/regex-quantifiers.html#helpful)]. This ensures that everything this is matched up until the first closing paren, ensuring that grouped matches are treated separately:

``` js
> joinedSelector.replace(/:matches\(.*?\)/g, function(substr) {
  console.log(substr)
  return substr.replace(/,/g, '&')
}).split(',')
:matches(:hover,:active,:focus)
:matches(:hover,:active,:focus)
[ '.foo:matches(:hover&:active&:focus)',
  '.baz:matches(:hover&:active&:focus)' ]
```

With the introduction of the `?` we can ensure that the grouped selectors aren't combined. I'm sure there's likely a more elegant way to achieve what I was intending to do. But I'm anything but a master of regex.

michelle replied:
Ah thanks for the reference--I didn't know that existed.

Michelle

On Wed, Feb 4, 2015 at 2:11 PM, John Otander notifications@github.com
wrote:

> In index.js
> https://github.com/michelle/rework-matches/pull/1#discussion_r24124860:
> 
> > @@ -34,9 +38,10 @@ module.exports = function() {
> > 
> > ```
> >    // Unfortunately rework unconditionally splits selectors by comma, so our
> >    // :matches(...) selector gets munged.
> > ```
> > -      var actualSelectors = rule.selectors.join(',').replace(/:matches(.*)/g, function(substr) {
> > -      var actualSelectors = rule.selectors.join(',').replace(/:matches(.*?)/g, function(substr) {
> 
> No worries. I very well could be doing it improperly.
> 
> My intention was that the regex will now only match up to the first
> closing paren. However, the .\* is greedy and will match everything
> between :matches( and the last ) of a string. So the regex would continue
> when there was a chained selector all the way to the last closing paren of
> the joined selectors:
> 
> $ node> var joinedSelector = '.foo:matches(:hover,:active,:focus),.baz:matches(:hover,:active,:focus)'undefined> joinedSelector.replace(/:matches(._)/g, function(substr) {
>   return substr.replace(/,/g, '&')
> }).split(',')
> [ '.foo:matches(:hover&:active&:focus)&.baz:matches(:hover&:active&:focus)' ]> joinedSelector.replace(/:matches(._)/g, function(substr) {
>   console.log(substr)
>   return substr.replace(/,/g, '&')
> }).split(','):matches(:hover,:active,:focus),.baz:matches(:hover,:active,:focus)
> [ '.foo:matches(:hover&:active&:focus)&.baz:matches(:hover&:active&:focus)' ]
> 
> So, the .\* was the right approach, up until the closing paren because it
> resulted in an & outside the :matches parens where a selector was
> originally grouped.
> 
> However, according to the article, I could introduce the minimal matcher,
> ?, to make the .\* "lazy but helpful" [reference
> http://www.rexegg.com/regex-quantifiers.html#helpful]. This ensures
> that everything this is matched up until the first closing paren, ensuring
> that grouped matches are treated separately:
> 
> > joinedSelector.replace(/:matches(.*?)/g, function(substr) {
> >   console.log(substr)
> >   return substr.replace(/,/g, '&')
> > }).split(','):matches(:hover,:active,:focus):matches(:hover,:active,:focus)
> > [ '.foo:matches(:hover&:active&:focus)',
> >   '.baz:matches(:hover&:active&:focus)' ]
> 
> With the introduction of the ? we can ensure that the grouped selectors
> aren't combined. I'm sure there's likely a more elegant way to achieve what
> I was intending to do. But I'm anything but a master of regex.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/michelle/rework-matches/pull/1/files#r24124860.

SlexAxton replied:
I'm not saying we should do this, but it could be nice across lots of the rework plugins to standardize paren-matching. I think it's loosely implemented a bunch of different places.

We could pull in something like: https://github.com/dfcreative/parenthesis -- which I'd suppose is the 'node'-y thing to do.

michelle replied:
Yeah, I think in the long term the goal should be to submit a PR to rework that parses `:matches` selectors correctly (currently the commas are screwing things up), but I haven't had time to dig into that.

johno replied:
Agreed. This is definitely something that could be solved in a more central, standardized location -- whether through a node library, a rework-plugin, or in rework itself.

Though, I'm inclined to think this is best solved uniformly at the rework level since it can already handle `:not`, `:nth-child`, etc. I think its wise for the long term goal to be adding `:matches` support (since it's part of the CSS4 spec). Though, I also think that being able to handle `:some-pseudo-selector(:with, :a, :comma, :delimited, :list)` is a rework concern. Having it end up as five different selectors is just silly :smirk_cat: .

Perhaps if I get some time this weekend I might have a dive into rework and the css module and see how complex it'd be to implement.

johno commented:
Thanks!
